kedro/io/core.py

```diff
@@ -871,3 +881,70 @@ def validate_on_forbidden_chars(**kwargs: Any) -> None:
             raise DatasetError(
                 f"Neither white-space nor semicolon are allowed in '{key}'."
             )
+
+
+_C = TypeVar("_C")
+
+
+@runtime_checkable
+class CatalogProtocol(Protocol[_C]):
+    _datasets: dict[str, AbstractDataset]
+
+    def __contains__(self, ds_name: str) -> bool:
+        """Check if a dataset is in the catalog."""
+        ...
+
+    @property
+    def config_resolver(self) -> CatalogConfigResolver:
+        """Return a copy of the datasets dictionary."""
+        ...
+
+    @classmethod
+    def from_config(cls, catalog: dict[str, dict[str, Any]] | None) -> _C:
+        """Create a ``KedroDataCatalog`` instance from configuration."""
+        ...
+
+    def _get_dataset(
```

Comment:
Is it necessary to add all methods implemented by `DataCatalog` to the protocol? I thought with a protocol it's also possible to not implement all defined methods and add others?

Reply:
I only added methods exposed to the framework components, so it is possible to use catalogs implementing this protocol with session, context, etc. So this is the bare minimum needed to make both catalogs work with the current implementation.

If I skip some of them, then we accept catalogs that do not implement some of the required methods, so they fail when used within the current framework implementation.

The same goes for adding methods — we can add them, but then all the catalogs should implement them, which is not needed for now.

Reply:
Ah right, that makes total sense. Thanks for clarifying!